language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,377 @@
+        voted_handle: Event::EventHandle<VotedEvent>,
+        ballot_approved_handle: Event::EventHandle<BallotApprovedEvent>,
+    }
+
+    /// A counter which is stored under the proposers address and gets incremented
+    /// everytime they create a new ballot. This is used for creating unique
+    /// global identifiers for Ballots
+    struct BallotCounter has key {
+        counter: u64,
+    }
+
+    /// CreateBallotEvent is emitted when a ballot is
+    /// created by a proposer
+    struct CreateBallotEvent<Proposal: store + drop> has drop, store {
+        ballot_id: BallotID,
+        ballot: Ballot<Proposal>,
+    }
+
+    /// RemoveBallotEvent is emitted when a ballot has
+    /// been removed. This can either happen because:
+    /// * ballot was approved
+    /// * ballot was manually removed by the proposer
+    /// * ballot was expired and garbage collected
+    struct RemoveBallotEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// VotedEvent is emitted when a valid vote has
+    /// been accepted by the ballot
+    struct VotedEvent has drop, store {
+        ballot_id: BallotID,
+        voter: address,
+        vote_weight: u64,
+    }
+
+    /// BallotApprovedEvent is emitted when a ballot has
+    /// been approved by the voters
+    struct BallotApprovedEvent has drop, store {
+        ballot_id: BallotID,
+    }
+
+    /// The maximum number of ballots allowed per proposal type
+    /// per address.
+    const MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS: u64 = 256;
+
+    /// The provided timestamp(s) were invalid
+    const EINVALID_TIMESTAMP: u64 = 1;
+    /// The address already contains has the maximum of ballots allowed
+    /// MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS
+    const ETOO_MANY_BALLOTS: u64 = 2;
+    /// Ballot with the provided id was not found
+    const EBALLOT_NOT_FOUND: u64 = 3;
+    /// Proposal details in the vote do not match the proposal details
+    /// in the ballot
+    const EBALLOT_PROPOSAL_MISMATCH: u64 = 4;
+    /// Voter not allowed to vote in the ballot
+    const EINVALID_VOTER: u64 = 5;
+    /// current timestamp > ballot expiration time
+    const EBALLOT_EXPIRED: u64 = 7;
+    /// Voter has already voted in this ballot
+    const EALREADY_VOTED: u64 = 8;
+
+    /// A constructor for BallotID
+    public fun new_ballot_id(
+        counter: u64,
+        proposer: address,
+    ): BallotID {
+        BallotID {
+            counter,
+            proposer,
+        }
+    }
+
+    /// A constructor for WeightedVoter
+    public fun new_weighted_voter(
+        weight: u64,
+        voter: vector<u8>,
+    ): WeightedVoter {
+        WeightedVoter {
+            weight,
+            voter,
+        }
+    }
+
+    /// Create a ballot under the signer's address and return the `BallotID`
+    public fun create_ballot<Proposal: store + copy + drop>(
+        ballot_account: &signer,
+        proposal: Proposal,
+        proposal_type: vector<u8>,
```

Comment:
I'm wondering if there's any way to restrict the `proposal_type` here to be a specific format, or to have to be in a list somewhere in order to prevent people from using useless/unrecognized proposal_type strings

Reply:
We can restrict it to lower alphanumeric characters plus underscores and maybe also restrict length.

Having a list is not great if we want this module to be used outside of DiemFramework by third party modules.

Reply:
I think if we don't have a central registry (and that isn't desired) then I think restricting the format of these won't buy much, so I'm fine keeping it as-is for now :)